File: ReviewJapanese/ReviewJapanese.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

using SplitOpts = System.StringSplitOptions;

public class ReviewJapanese
{
    internal record JPWord
    {
        public string Kana { get; init; }
        public string Kanji { get; init; }
        public string Definition { get; init; }
    };

    static int Main(string[] args)
    {
        if (args.Length < 1)
        {
            Console.WriteLine("Apologies, but the japanese dictionary file is"
                              + " needed to populate the question pool :(");
            return -1;
        }

        Console.OutputEncoding = Encoding.UTF8;
        List<string> wordsPool = GetWordsFromOrgDict(args[0]);
        return 0;
    }

    private static List<string> GetWordsFromOrgDict(string jpOrgFile)
    {
        List<string> words = new List<string>();

        foreach (string line in File.ReadLines(jpOrgFile, Encoding.UTF8))
        {
            if (!line.StartsWith('-')) continue;

            string[] parts = line.Split(' ', SplitOpts.RemoveEmptyEntries
                                             | SplitOpts.TrimEntries);

            // Parts[0] = The dash denoting the list item
            // Parts[1] = Hiragana/Katakana writing
            // Parts[2] = Kanji writing
            // Parts[3..] = Definitions

            string kanaWriting = parts[1];
            string kanjiWriting = parts[2].Trim('(', ')', ':');
            string definitions = string.Join(' ', parts[3..]);

            words.Add(new JPWord(kanaWriting, kanjiWriting, definitions));
        }

        return words;
    }
}

File: reviewjapanese.jl
#!/usr/bin/julia

using Random
using StatsBase # Needs to be installed via the Julia REPL with Pkg.add("StatsBase")

"""
"""
struct JPWord
    kana::String
    kanji::String
    meanings::String
end

"""
"""
function getparams()
    wordfile = ""
    numquestions = 0

    if length(ARGS) < 2
        print("\n")

        if length(ARGS) < 1
            print("Enter the file where the word list is: ")
            wordfile = readline(stdin)
        else
            wordfile = ARGS[1]
        end

        print("Enter the number of questions you want to review: ")
        numquestions = parse(Int, readline(stdin))

    else
        wordfile = ARGS[1]
        numquestions = parse(Int, ARGS[2])

    end

    return wordfile, numquestions
end

"""
"""
function getwords(jporgfile::String)
    words = Vector{JPWord}()

    for line in eachline(jporgfile)
        if ! startswith(line, "-") continue end

        parts = split(line, " ")

        # Parts[1] = The dash denoting the list item
        # Parts[2] = Hiragana/Katakana writing
        # Parts[3] = Kanji writing
        # Parts[4..] = Definitions

        kanawriting = parts[2]
        kanjiwriting = strip(parts[3], ['(', ')', ':'])
        meanings = join(parts[4:length(parts)], " ")

        push!(words, JPWord(kanawriting, kanjiwriting, meanings))
    end

    return words
end

"""
"""
function run_kanji2englishtest(wordlist::Vector{JPWord})
    numcorrects = 0
    misses = Vector{@NamedTuple{q::JPWord, a::String}}()
    total = length(wordlist)

    shuffle!(wordlist)

    for question in wordlist
        println("\nWhat does this Kanji $(question.kanji) mean?")
        answer = readline(stdin)

        if ! isempty(answer) && occursin(answer, lowercase(question.meanings))
            numcorrects += 1
        else
            push!(misses, (q=question, a=answer))
        end
    end

    print("\nYou had $numcorrects/$total correct answers")

    if numcorrects == total
        print("!\n")
        println("Way to go! You got all questions of this session right!")
    else
        print(".\n")
        println("You missed $(total - numcorrects) questions :(")
        println("\nHere are the missed questions, so you can review them:\n")

        for miss in misses
            println("- Kanji:    $(miss.q.kanji)")
            println("  Kana:     $(miss.q.kana)")
            println("  Meanings: $(miss.q.meanings)\n")
            println("You answered: '$(miss.a)'\n")
        end
    end
end

"""
"""
function main()
    wordfile, numquestions = getparams()

    wordpool = getwords(wordfile)
    questions = sample(wordpool, numquestions)
    run_kanji2englishtest(questions)

    return 0
end

main()
